File: examples/full/RemonEx-objc/SimpleCallViewController.m
//
//  SimpleCallViewController.m
//  verySimpleRomen-objc
//
//  Created by lhs on 2018. 9. 11..
//  Copyright © 2018년 Remote Monster. All rights reserved.
//

#import "SimpleCallViewController.h"

@interface SimpleCallViewController ()
@property (weak, nonatomic) IBOutlet UILabel *ChannelIdLabel;
@property (weak, nonatomic) IBOutlet UITextField *ChannelIdField;
@property (weak, nonatomic) IBOutlet UIView *boxView;
@property (weak, nonatomic) IBOutlet UILabel *statusLabel;
@property (weak, nonatomic) IBOutlet UIButton *closeButton;

@property (strong, nonatomic) IBOutlet RemonCall *remonCall;
@end

BOOL muted = NO;
@implementation SimpleCallViewController

- (IBAction)volumeRatioP:(id)sender {
    [self.remonCall setVolumeRatio:self.remonCall.volumeRatio + 0.1];
}

- (IBAction)volumeRatioM:(id)sender {
    [self.remonCall setVolumeRatio:self.remonCall.volumeRatio - 0.1];
}

- (IBAction)muteLocalAudio:(id)sender {
    [self.remonCall muteLocalAudioWithMute:!muted];
    muted = !muted;
}

- (IBAction)connectChannel:(id)sender {
    NSString *chId = self.ChannelIdField.text;
    if (chId == nil || chId.length == 0) {
        NSInteger rand = arc4random_uniform(9999);
        chId = [NSString stringWithFormat:@"%ld", (long)rand];
    }
    
    [self.remonCall connect:chId :self.customConfig];
}

- (IBAction)closeRemon:(id)sender {
    [self.remonCall closeRemon];
    
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    
    [self.remonCall onInitWithBlock:^{
        [self.statusLabel setText:@"INIT"];
    }];
    
    [self.remonCall onConnectWithBlock:^(NSString * _Nullable chId) {
        [self.statusLabel setText:@"CONNECT"];
        if (chId != nil) {
            [self.closeButton setEnabled:YES];
            [self.ChannelIdLabel setText:chId];
        } else {
            [self.ChannelIdLabel setText:@"connection failed"];
        }
    }];
    
    [self.remonCall onCompleteWithBlock:^{
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.statusLabel setText:@"COMPLETE"];
            [self.closeButton setEnabled:YES];
            [self.ChannelIdLabel setText:self.remonCall.channelID];
            [self.ChannelIdField setText:@""];
            [self.boxView setHidden:YES];
        });
        
        //덤프 기록 시작
//        [self.remonCall startDumpWithFileName:@"audio.aecdump" maxSizeInBytes:1000 * 1024* 1024];
    }];
    
    [self.remonCall onCloseWithBlock:^(RemonCloseType type){
        NSLog(@"zzzz %ld" , type);
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.boxView setHidden:NO];
            [self.closeButton setEnabled:NO];
            [self.ChannelIdLabel setText:@"connection closed"];
            [self.statusLabel setText:@"CLOSE"];
        });
        
        //덤프 기록 중지
        //onClose()에서 하셔도 되고, 별도의 이벤트에서 하셔도 됩니다.
        [self.remonCall stopDump];
        
        //덤프 파일을 m4a 형식으로 인코딩 하는 메소드 입니다.
        //이 과정은 시간이 소요 됩니다. onClose()에서 하셔도 되고, 덤프 파일 위치만 알고 있다면 별도의 이벤트에서 하셔도 됩니다.
//        [RemonCall unpackAecDumpWithDumpName:@"audio.aecdump" resultFileName:@"unpack.m4a" progress:^(NSError * _Nullable erro, enum REMON_AECUNPACK_STATE state) {
//        }];
        
//        [RemonCall unpackAecDumpWithDumpName:@"audio.aecdump" resultFileName:@"unpack.m4a" avPreset:REMON_AECUNPACK_PRESETMP4MEDIUM progress:^(NSError * _Nullable erro, enum REMON_AECUNPACK_STATE state) {
//            
//        }];
    }];
    
    [self.remonCall onRemonStatReportWithBlock:^(RemonStatReport * _Nonnull stat) {
        RatingValue *remonRating = [stat getHealthRating];
    }];
    
    [self.remonCall onRemoteVideoSizeChangedWithBlock:^(UIView * _Nullable view, CGSize size) {
        CGFloat videoHeight = size.height;
        CGFloat videoWidth = size.width;
        CGFloat videoRatio = videoWidth / videoHeight;
        
        CGFloat myViewWidth = view.frame.size.width;
        CGFloat myViewHeight = view.frame.size.height;
        CGFloat myViewRatio = myViewWidth / myViewHeight;
        
        if (videoRatio < 1.0) { // 방송 영상이 세로입니다.
            if (myViewRatio < 1.0) { // 시청자 뷰가 세로 입니다.
                CGFloat computedWidth = myViewHeight * videoRatio;
                dispatch_async(dispatch_get_main_queue(), ^{
                    view.frame = CGRectMake(0.0, 0.0, computedWidth, myViewHeight);
                    view.center = self.view.center;
                });
            } else {
                //                    NOOP
            }
        } else {
            //                NOOP
        }
    }];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    [self.remonCall closeRemon];
}
- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


@end

File: examples/full/RemonFull/SimpleCallViewController.swift
//
//  SimpleCallViewController.swift
//  verySimpleRemon
//
//  Created by hyounsiklee on 2018. 4. 27..
//  Copyright © 2018년 Remote Monster. All rights reserved.
//

import UIKit
import RemoteMonster

class SimpleCallViewController: UIViewController {
    
    @IBOutlet var remonCall: RemonCall!
    @IBOutlet weak var boxView: UIView!
    @IBOutlet weak var chField: UITextField!
    @IBOutlet weak var chLabel: UILabel!
    
    var customConfig:RemonConfig?
    var muted = false
    
    @IBAction func volumeRatioP(_ sender: Any) {
        self.remonCall.volumeRatio = self.remonCall.volumeRatio + 0.1
    }
    
    @IBAction func volumeRatioM(_ sender: Any) {
        self.remonCall.volumeRatio = self.remonCall.volumeRatio - 0.1
    }
    
    @IBAction func ewf(_ sender: Any) {
        self.remonCall.showLocalVideoStat = true
        self.remonCall.showRemoteVideoStat = true
    }
    
    @IBAction func awedawef(_ sender: Any) {
        self.remonCall.muteLocalAudio(mute: !self.muted)
        self.muted = !self.muted
    }
    
    @IBAction func touchConnectButton(_ sender: Any) {
        self.view.endEditing(true)
        let chid = self.chField.text
        if chid != nil && chid!.count > 0 {
            //config is nilable
            remonCall.connect(chid!, customConfig)
            self.chLabel.text = chid
        } else {
            let rand:String! = String(arc4random_uniform(9999))
            //config is nilable
            remonCall.connect(rand, customConfig)
            self.chLabel.text = rand
        }
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
//        let apapap = aecunpack()
//        apapap.run(dumpName: "dump.dump", resultFileName: "result.m4a") { (err, state) in
//            if let err = err {
//                print(err.localizedDescription)
//            } else {
//                
//            }
//        }
        
        do {
            try AVAudioSession.sharedInstance().setCategory(AVAudioSessionCategoryPlayback, mode: AVAudioSessionModeDefault)
            try AVAudioSession.sharedInstance().setActive(true, with: [])
        } catch {
            print(error)
        }
        
        remonCall.useFrontCamera = false
        remonCall.onInit {
            DispatchQueue.main.async {
                self.boxView.isHidden = true
            }
        }
        
        remonCall.onComplete { () in
            DispatchQueue.main.async {
                self.chLabel.text = self.remonCall.channelID
            }
        }
        
        remonCall.onRemoteVideoSizeChanged { (remoteView, size) in
            print("aaaaa", size)
            let newFrame = CGRect(x: 0.0, y: 0.0, width: size.width, height: size.height)
            remoteView?.frame = newFrame
        }
        
        remonCall.onLocalVideoSizeChanged { (localView, size) in
            print("aaaaa", size, "bbb")
        }
        
        remonCall.onClose { (_) in
            DispatchQueue.main.async {
//                self.chLabel.text = "Close Remon-   "
            }
        }
        
        remonCall.onConnect { (ss) in
            
        }
        
        remonCall.onError { (error) in
            DispatchQueue.main.async {
                self.chLabel.text = error.localizedDescription
            }
        }
        
        remonCall.onRetry { (completed) in
            print(completed)
        }
        
        remonCall.onRemoteVideoSizeChanged { (view, size) in
            print("Debug onRemoteVideoSizeChanged", size)
            print("Debug self.remonCall\(self.remonCall.remoteView.hashValue) and view\(view.hashValue) is same")
            
            let videoHeight = size.height
            let videoWidth = size.width
            let videoRatio:CGFloat = videoWidth / videoHeight
            
            guard let myView = view else { return }
            
            let myViewWidth:CGFloat = myView.frame.size.width
            let myViewHeight:CGFloat = myView.frame.size.height
            let myViewRatio:CGFloat = myViewWidth / myViewHeight
            
            if videoRatio < 1.0 {
                if myViewRatio < 1.0 {
                    let computedWidth:CGFloat = myViewHeight * videoRatio
                    print("Debug computedWidth", computedWidth)
                    DispatchQueue.main.async {
                        myView.frame = CGRect(x: 0.0, y: 0.0, width: computedWidth, height: myViewHeight)
                        myView.center = self.view.center
                    }
                } else {
//                    NOOP
                }
            } else {
//                NOOP
            }
        }
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        self.remonCall.closeRemon()
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
}
